File: Retos/Reto #19 - ANÁLISIS DE TEXTO [Media]/kotlin/mjordanaam.kt
/*
 * Crea un programa que analice texto y obtenga:
 * - Número total de palabras.
 * - Longitud media de las palabras.
 * - Número de oraciones del texto (cada vez que aparecen un punto).
 * - Encuentre la palabra más larga.
 *
 * Todo esto utilizando un único bucle.
 */

fun analyzeText(text: String) {
    var sentences: Int = 0
    var words = arrayOf<String>()
    var aux: String = ""
    var nwords: Int = 0
    var length: Int = 0
    var longest = hashMapOf<Int, Array<String>>(0 to arrayOf<String>())
    
    for (letter in text) {
        val regex1 = Regex("[.?!]+")
        val regex2 = Regex("[¿|:()/@%$,]+")
        if (regex1.matches(letter.toString())) {
            sentences++
            words += aux
            length += aux.length
            nwords++
            
            if (aux.length > longest.keys.first()) {
                longest = hashMapOf(aux.length to arrayOf<String>(aux))
            } else if (aux.length == longest.keys.first()) {           
                longest[aux.length] = longest.getOrDefault(aux.length, emptyArray()) + aux
            }
            
            aux = ""
        }else if (letter == ' ') {
            if (aux != "") {
                words += aux
                length += aux.length
                nwords++

                if (aux.length > longest.keys.first()) {
                    longest = hashMapOf(aux.length to arrayOf(aux))
                } else if (aux.length == longest.keys.first()) {
                    longest[aux.length] = longest.getOrDefault(aux.length, emptyArray()) + aux
                }
                aux = ""
            }
        }else if (regex2.matches(letter.toString())){
            
        }else {
            aux += letter
        }
    }
    
    var average = length/nwords
    
    println("> Total Words = $nwords")
    println("> Average length = $average")
    println("> Number of sentences = $sentences")
    println("> Longest word/words (${longest.keys.first()} letters) = ${longest.get(longest.keys.first()).contentToString()}")
    println()

}

fun main() {
    var text: String = "Hello. World!"
    
    analyzeText(text)


    text = "Crea un programa que analice texto y obtenga:." + 
            "Número total de palabras." +
            "Longitud media de las palabras." +
            "Número de oraciones del texto (cada vez que aparecen un punto)." +
            "Encuentre la palabra más larga." + 
            "Todo esto utilizando un único bucle."
    
    analyzeText(text)

    text = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis condimentum mauris non enim mollis dictum. In " +
           "feugiat ut justo in vulputate. Aliquam erat volutpat. Vivamus porttitor commodo felis, sed gravida eros " +
           "fermentum non. Mauris sagittis id neque sit amet ullamcorper. Sed eu pretium ex, ut ornare quam. Maecenas " +
           "consectetur elit a nisi maximus, et vehicula lorem finibus. Duis sapien justo, placerat in vestibulum a, " +
           "vestibulum non lacus. Sed egestas, nisi."
    
    analyzeText(text)

}

File: Retos/Reto #19 - ANÁLISIS DE TEXTO [Media]/python/mjordanaam.py
"""
/*
 * Crea un programa que analice texto y obtenga:
 * - Número total de palabras.
 * - Longitud media de las palabras.
 * - Número de oraciones del texto (cada vez que aparecen un punto).
 * - Encuentre la palabra más larga.
 *
 * Todo esto utilizando un único bucle.
 */
"""
import re

def analyze_text(text: str):
    words = []
    sentences = 0
    longest = {0: ""}
    aux = ""
    nwords = 0
    length = 0

    for letter in text:
        if re.search(r"[?.!]+", letter):
            sentences += 1
            words.append(aux)
            length += len(aux)
            nwords += 1
            
            if len(aux) > list(longest.keys())[0]:
                longest = {
                    len(aux): [aux]
                }
            elif len(aux) == list(longest.keys())[0]:
                longest[len(aux)].append(aux)
            aux = ""
        elif letter == " ":
            if aux != "":
                words.append(aux)
                length += len(aux)
                nwords += 1
                
                if len(aux) > list(longest.keys())[0]:
                    longest = {
                        len(aux): [aux]
                    }
                elif len(aux) == list(longest.keys())[0]:
                    longest[len(aux)].append(aux)
                else:
                    pass
                
                aux = ""
        elif re.search(r"[¿|:()/@%$,]+", letter):
            pass
        else:
            aux = aux + letter
    
    average = round(length/nwords, 2)

    print(f"> Total Words = {nwords}")
    print(f"> Average length = {average}")
    print(f"> Number of senteces = {sentences}")
    print(f"> Longest word/words ({list(longest.keys())[0]} letters) = {list(longest.values())[0]}")
    print()
  

text = (
    "Hello."
    "World!"
)
analyze_text(text)

text = (
    "Crea un programa que analice texto y obtenga:."
    "Número total de palabras."
    "Longitud media de las palabras."
    "Número de oraciones del texto (cada vez que aparecen un punto)."
    "Encuentre la palabra más larga."
    "Todo esto utilizando un único bucle."
)
analyze_text(text)

text = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis condimentum mauris non enim mollis dictum. In " \
       "feugiat ut justo in vulputate. Aliquam erat volutpat. Vivamus porttitor commodo felis, sed gravida eros " \
       "fermentum non. Mauris sagittis id neque sit amet ullamcorper. Sed eu pretium ex, ut ornare quam. Maecenas " \
       "consectetur elit a nisi maximus, et vehicula lorem finibus. Duis sapien justo, placerat in vestibulum a, " \
       "vestibulum non lacus. Sed egestas, nisi."

analyze_text(text)